Sources/NIOHTTPClient/HTTPHandler.swift

```diff
@@ -18,25 +18,39 @@ import NIOConcurrencyHelpers
 import NIOHTTP1
 import NIOSSL
 
-public extension HTTPClient {
-    enum Body: Equatable {
-        case byteBuffer(ByteBuffer)
-        case data(Data)
-        case string(String)
-
-        var length: Int {
-            switch self {
-            case .byteBuffer(let buffer):
-                return buffer.readableBytes
-            case .data(let data):
-                return data.count
-            case .string(let string):
-                return string.utf8.count
+extension HTTPClient {
+    public struct Body {
+        public var length: Int?
+        public var provider: (@escaping (IOData) -> EventLoopFuture<Void>) -> EventLoopFuture<Void>
```

Comment:
Using a struct here would give us more flexibility to add / deprecate methods in the future. It would also allow users of the package to add conformances / extensions. 

Something like `Body.StreamWriter` or `Body.Provider`:

```swift
extension HTTPClient.Body 
    var writer: @escaping (StreamWriter) -> EventLoopFuture<Void>

    public struct StreamWriter {
        let closure: @escaping (IOData) -> EventLoopFuture<Void>
        public func write(_ data: IOData) -> EventLoopFuture<Void> {
            return self.closure(data)
        }
    } 

    public static func stream(length: Int? = nil, _ writer: (StreamWriter) -> EventLoopFuture<Void>) -> 
       ...
    }
}
...
```

This also reduces the mental load to understand the code a bit. Here's example usage:

```swift
let body: HTTPClient.Body = .stream(length: 50) { stream in
    var request = try! Request(url: "http://localhost:\(httpBin.port)/events/10/1")
    request.headers.add(name: "Accept", value: "text/event-stream")

    let delegate = CopyingDelegate { part in
        stream.write(.byteBuffer(part))
    }
    return httpClient.execute(request: request, delegate: delegate).future
}
```

I'd also make the `Body.provider` property private if possible

Reply:
good idea, refactored, thanks!